### ISSUE TITLE
Drag & Drop to Menubar

### ISSUE THREAD

**Issue by latenitefilms** [Hammerspoon/hammerspoon#1057]
@asmagill - have you ever played around with NSDragPboard?

Is there any way of getting the contents of something from an random application (say Final Cut Pro), if it's dropped on a menubar icon for example?

**Comment by cmsj** [comment]
Somewhat related, I always thought it might be nice to have some mechanism for receiving drags, a la Yoink.

**Comment by latenitefilms** [comment]
I've never heard of Yoink before - that looks awesome! This would definitely be something cool to add to Hammerspoon!

**Comment by cmsj** [comment]
FWIW it's quite important to me that we don't specifically target other Mac utilities to clone. I don't want to hurt indie developers.
I'd rather that we add generic functionality that can be used to build whatever kind of utility features our users want to make.

**Comment by latenitefilms** [comment]
Absolutely! I only want it so that I can "accept" things dragged from Final Cut Pro (something Yoink can't currently accept for some reason).

**Comment by asmagill** [comment]
@cmsj, I tend to agree -- while I don't think I've really thought about it in those terms until now, what I think I've done in the past is add basic module functionality at the `hs.module` level and then maybe include an example of it's use in my personal config (also available via github) mimic'ing other tools... I will definitely make a point to continue that model going forward now that you've stated it in a clearer way then I ever did.

@latenitfilms, among other items that can be captured on the clipboard is a file URL... if you select a document in the finder and "Copy" it (i.e. Cmd-C (which can also be mimiced through the eventtap module) or by selecting Copy from the edit or a pop-up menu), one of the items placed on the clipboard is a file URL which points to the "copied" file... I've tested this much.  Following that logic, it may also be possible to set a file URL in the pasteboard with `hs.pasteboard.writeObjects({ url = "file://...full..path..to..file" })` and "Paste" it into a receiving application... haven't tested that part, though.

And I know that an eventtap watcher (`hs.eventtap.new`) can receive drag events... I don't know of anyone who has tracked or played with them enough to know what else we might need to add to actually be able to take advantage of tracking these events, though.  I'm working on some additions to `hs.eventtap` which should allow us to create arbitrary events above and beyond the mouseUp/Down and keyUp/Down which can already be created, so this may also lead to us being able to perform drags under programatic control, but it's too early to say if my additions will be sufficient or if, perhaps like tracking them, more additions might be of use... it's fertile ground and if you get a chance to play with it some more, let us know if specific additions might help.

**Comment by latenitefilms** [comment]
Thanks for the detailed reply @asmagill !

I've just had a quick play with **hs.eventtap.event.types.leftMouseDragged**, and it looks like I can definitely "trigger" things when something is being dragged - but how do I get the contents of WHAT'S being dragged?

**Comment by latenitefilms** [comment]
@asmagill - I was just wondering if you've had a chance to play around with "dragging and dropping" by any chance? I'd still LOVE to work out if it's possible to get a **hs.menubar** to "accept" items dragged into it. Thoughts?

**Comment by latenitefilms** [comment]
Also... if it's technically not possible for a **hs.menubar** to "accept" things dragged into it - maybe a workaround would be to just allow the Hammerspoon dock icon to "accept" things?

**Comment by asmagill** [comment]
I have not looked into this yet... My suspicion is that it will probably require less effort to enable dragging to the dock icon, but I really don't know. It is doubtful that I will have the time to work on this in the near future, but I do think that its a good suggestion for an extension if someone is looking for the chance to contribute.

**Comment by latenitefilms** [comment]
@asmagill - No worries! I'll try do some digging and research. I might need some help from yourself or @cmsj though, to translate what will no-doubt be Objective-C to Lua code - which is above my pay grade!

**Comment by latenitefilms** [comment]
@asmagill @cmsj - Is [this](http://stackoverflow.com/a/26810727/6925202) helpful at all?

**Comment by latenitefilms** [comment]
FYI: My workaround for now is to just have a user selectable folder as my "drop zone", so for example, you can just create a folder called "Hammerspoon Dropbox" on your desktop, put it in the dock, then use that to "accept" things dragged and dropped from FCPX. However, eventually I'd still LOVE to be able to drag these things to the **hs.menubar** instead (or if that proves too technically challenging, the Hammerspoon Dock). I'll keep doing some research into this, and see what I can come up with...

**Comment by latenitefilms** [comment]
I've just come [across a project on Github](https://github.com/qiushihe/emacs-launcher) that also offers drag and drop support to the menubar. It looks like it's in Swift, so I'm not sure if it's useful or not?

**Comment by latenitefilms** [comment]
@cmsj - Seeing as you're currently online, I was just wondering if you have any ideas or pointers in regards to adding drag/drop support to `hs.menubar`? I'd had a play with trying to modify `hs.menubar` to make it work, but I know very little about Objective-C, so not having much luck. Is there anyone else in the Hammerspoon community that might be interested in tackling this problem?

**Comment by cmsj** [comment]
It looks to me like supporting drag&drop in menubar items requires using a custom view, which we don't currently do. I think in the fullness of time we're going to need to move to a custom view, since that seems to be the way the wind is blowing in Apple's SDK, but I don't think it'll make 0.9.53.

It shouldn't be too hard to support drag&drop to hs.canvas/hs.drawing objects though (but probably not in the kind of 0.9.53 timeframe I'm hoping for).

**Comment by asmagill** [comment]
Speaking of hs.menubar changes, when we do move the menubar portion to the use of an NSView (NSButton), I'd like to also consider breaking out the actual menu portion of it into a separate userdata type as well... this would facilitate a common language and interface for creating menus that could be used in other places -- in toolbar buttons, when clicking on the title of a webview/console title, in some of the other views I hope to eventually add to canvas, etc.

@cmsj, re drag-and-drop for canvas... I doubt it, but if I have time this weekend, I'll look into it.  However it's most likely a 0.9.54 thing.

**Comment by latenitefilms** [comment]
> Speaking of hs.menubar changes, when we do move the menubar portion to the use of an NSView (NSButton), I'd like to also consider breaking out the actual menu portion of it into a separate userdata type as well... this would facilitate a common language and interface for creating menus that could be used in other places -- in toolbar buttons, when clicking on the title of a webview/console title, in some of the other views I hope to eventually add to canvas, etc.

I'm sure I've already discussed this in another issue, but another awesome thing to add to `hs.menubar` would be the ability to have a `hs.webview` "attached" to the menubar. That way we could design our own custom menubar interfaces - and build things like the Dropbox menubar men for example - as opposed to just a boring text-based menu list.

**Comment by cmsj** [comment]
I'm going to close this out - it's yet another issue that is blocked on us rewriting hs.menubar. Let's wait and see if that ever happens. If it does, we can think about things like this.

**Comment by asmagill** [comment]
FYI menubar rewrite in progress; will add with guitk in upcoming pull (needs some more testing and documentation)
